File: src/mnemonic/converter/midi.py
"""MIDI変換モジュール

MIDIファイルをAndroid互換のOGG Vorbis形式に変換するための機能を提供する。
FluidSynthを使用してMIDIをWAVにレンダリングし、FFmpegでOGGに変換する。
"""

import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path

from .base import BaseConverter, ConversionResult, ConversionStatus


@dataclass(frozen=True)
class MidiConversionConfig:
    """MIDI変換設定を表すデータクラス

    MIDI変換処理のパラメータを保持する不変データクラス。

    Attributes:
        soundfont_path: サウンドフォントファイルのパス
        sample_rate: 出力サンプルレート（Hz）
        audio_codec: 使用する音声コーデック
        audio_quality: 音声品質（0-10、VBR用）
    """

    soundfont_path: Path
    sample_rate: int
    audio_codec: str
    audio_quality: int


class MidiConverter(BaseConverter):
    """MIDIファイルをOGG Vorbis形式に変換するConverter

    FluidSynthを使用してMIDIをWAVにレンダリングし、
    FFmpegでOGG Vorbis形式に変換する。

    変換フロー:
    1. FluidSynth + SoundFont で MIDI → WAV
    2. FFmpeg で WAV → OGG Vorbis
    """

    DEFAULT_SOUNDFONT_PATH = Path("/usr/share/sounds/sf2/FluidR3_GM.sf2")

    def __init__(
        self,
        soundfont_path: Path | None = None,
        sample_rate: int = 44100,
        audio_codec: str = "libvorbis",
        audio_quality: int = 4,
        timeout: int = 300,
    ) -> None:
        """MidiConverterを初期化する

        Args:
            soundfont_path: サウンドフォントファイルのパス（デフォルト: FluidR3_GM.sf2）
            sample_rate: 出力サンプルレート（デフォルト: 44100 Hz）
            audio_codec: 使用する音声コーデック（デフォルト: libvorbis）
            audio_quality: 音声品質（デフォルト: 4、VBR用）
            timeout: 変換処理のタイムアウト秒数（デフォルト: 300秒）
        """
        self._soundfont_path = soundfont_path or self.DEFAULT_SOUNDFONT_PATH
        self._sample_rate = sample_rate
        self._audio_codec = audio_codec
        self._audio_quality = audio_quality
        self._timeout = timeout

    @property
    def supported_extensions(self) -> tuple[str, ...]:
        """対応する拡張子のタプルを返す

        Returns:
            対応するMIDIファイル拡張子のタプル
        """
        return (".mid", ".midi")

    def can_convert(self, file_path: Path) -> bool:
        """このConverterで変換可能なファイルかを判定する

        Args:
            file_path: 判定対象のファイルパス

        Returns:
            変換可能な場合True、そうでない場合False
        """
        return file_path.suffix.lower() in self.supported_extensions

    def convert(self, source: Path, dest: Path) -> ConversionResult:
        """MIDIファイルをOGG Vorbis形式に変換する

        Args:
            source: 変換元MIDIファイルのパス
            dest: 変換先ファイルのパス

        Returns:
            変換結果を表すConversionResultオブジェクト
        """
        if not source.exists():
            return ConversionResult(
                source_path=source,
                dest_path=None,
                status=ConversionStatus.FAILED,
                message=f"変換元ファイルが見つかりません: {source}",
            )

        if not self._soundfont_path.exists():
            return ConversionResult(
                source_path=source,
                dest_path=None,
                status=ConversionStatus.FAILED,
                message=f"サウンドフォントが見つかりません: {self._soundfont_path}",
            )

        bytes_before = self._get_file_size(source)

        dest.parent.mkdir(parents=True, exist_ok=True)

        try:
            # 一時WAVファイルを作成
            with tempfile.NamedTemporaryFile(suffix=".wav", delete=False) as tmp_wav:
                tmp_wav_path = Path(tmp_wav.name)

            try:
                # Step 1: FluidSynth で MIDI → WAV
                fluidsynth_result = self._run_fluidsynth(source, tmp_wav_path)
                if fluidsynth_result is not None:
                    return fluidsynth_result

                # Step 2: FFmpeg で WAV → OGG
                ffmpeg_result = self._run_ffmpeg(tmp_wav_path, dest)
                if ffmpeg_result is not None:
                    return ffmpeg_result

            finally:
                # 一時ファイルをクリーンアップ
                if tmp_wav_path.exists():
                    tmp_wav_path.unlink()

            bytes_after = self._get_file_size(dest)

            return ConversionResult(
                source_path=source,
                dest_path=dest,
                status=ConversionStatus.SUCCESS,
                bytes_before=bytes_before,
                bytes_after=bytes_after,
            )

        except Exception as e:
            return ConversionResult(
                source_path=source,
                dest_path=None,
                status=ConversionStatus.FAILED,
                message=f"MIDI変換に失敗しました: {e}",
            )

    def _run_fluidsynth(self, source: Path, wav_output: Path) -> ConversionResult | None:
        """FluidSynthを実行してMIDIをWAVに変換する

        Args:
            source: 入力MIDIファイルのパス
            wav_output: 出力WAVファイルのパス

        Returns:
            エラー時はConversionResult、成功時はNone
        """
        cmd = [
            "fluidsynth",
            "-ni",  # No interactive mode, no shell
            "-g",
            "1.0",  # Gain
            "-r",
            str(self._sample_rate),  # Sample rate
            "-F",
            str(wav_output),  # Output file
            str(self._soundfont_path),
            str(source),
        ]

        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self._timeout,
            )

            if result.returncode != 0:
                return ConversionResult(
                    source_path=source,
                    dest_path=None,
                    status=ConversionStatus.FAILED,
                    message=f"FluidSynth変換に失敗しました: {result.stderr}",
                )

            return None

        except subprocess.TimeoutExpired:
            return ConversionResult(
                source_path=source,
                dest_path=None,
                status=ConversionStatus.FAILED,
                message=f"FluidSynth処理がタイムアウトしました（{self._timeout}秒）",
            )
        except FileNotFoundError:
            return ConversionResult(
                source_path=source,
                dest_path=None,
                status=ConversionStatus.FAILED,
                message="FluidSynthが見つかりません。インストールしてください。",
            )

    def _run_ffmpeg(self, wav_input: Path, ogg_output: Path) -> ConversionResult | None:
        """FFmpegを実行してWAVをOGGに変換する

        Args:
            wav_input: 入力WAVファイルのパス
            ogg_output: 出力OGGファイルのパス

        Returns:
            エラー時はConversionResult、成功時はNone
        """
        cmd = [
            "ffmpeg",
            "-y",  # Overwrite output
            "-i",
            str(wav_input),
            "-c:a",
            self._audio_codec,
            "-q:a",
            str(self._audio_quality),
            str(ogg_output),
        ]

        try:
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=self._timeout,
            )

            if result.returncode != 0:
                return ConversionResult(
                    source_path=wav_input,
                    dest_path=None,
                    status=ConversionStatus.FAILED,
                    message=f"FFmpeg変換に失敗しました: {result.stderr}",
                )

            return None

        except subprocess.TimeoutExpired:
            return ConversionResult(
                source_path=wav_input,
                dest_path=None,
                status=ConversionStatus.FAILED,
                message=f"FFmpeg処理がタイムアウトしました（{self._timeout}秒）",
            )
        except FileNotFoundError:
            return ConversionResult(
                source_path=wav_input,
                dest_path=None,
                status=ConversionStatus.FAILED,
                message="FFmpegが見つかりません。インストールしてください。",
            )

    def is_fluidsynth_available(self) -> bool:
        """FluidSynthが利用可能かを確認する

        Returns:
            FluidSynthが利用可能な場合True、そうでない場合False
        """
        try:
            subprocess.run(
                ["fluidsynth", "--version"],
                capture_output=True,
                check=True,
            )
            return True
        except (FileNotFoundError, subprocess.SubprocessError):
            return False

File: internal/converter/midi.go
package converter

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// MuseScoreSoundfontPath / FluidR3SoundfontPath はGetDefaultSoundfontPathが
// 探索するデフォルトサウンドフォントのパス。
//
// why: テストで存在判定を差し替えられるよう、パッケージ変数として公開する
// （テストは一時的な差し替えと復元をdeferで行う）。
var (
	MuseScoreSoundfontPath = "/usr/share/sounds/sf3/MuseScore_General.sf3"
	FluidR3SoundfontPath   = "/usr/share/sounds/sf2/FluidR3_GM.sf2"
)

// GetDefaultSoundfontPath は利用可能なデフォルトサウンドフォントのパスを返す。
// MuseScoreSoundfontPathが存在すればそれを、そうでなければFluidR3SoundfontPathを
// 返す（存在確認自体は行わないFluidR3SoundfontPathへの最終フォールバック）。
func GetDefaultSoundfontPath() string {
	if _, err := os.Stat(MuseScoreSoundfontPath); err == nil {
		return MuseScoreSoundfontPath
	}

	return FluidR3SoundfontPath
}

// MidiConverter はFluidSynth+FFmpegを使いMIDIファイルをOGG Vorbis形式に
// 変換するConverter。
//
// 変換フロー:
//  1. FluidSynth + SoundFont で MIDI → WAV
//  2. FFmpeg で WAV → OGG Vorbis
type MidiConverter struct {
	soundfontPath string
	sampleRate    int
	audioCodec    string
	audioQuality  int
	timeout       time.Duration
	runner        CommandRunner
}

// NewMidiConverter はMidiConverterを初期化する。
// soundfontPathが空文字列の場合はGetDefaultSoundfontPath()を使用する。
// sampleRate/audioQualityが0以下の場合、audioCodecが空文字列の場合、
// timeoutが0以下の場合はデフォルト値
// (sampleRate=44100, audioCodec="libvorbis", audioQuality=4, timeout=300秒)を
// 使用する。runnerがnilの場合はos/execベースの既定実装を使用する。
//
// why not: audioQuality=0（ffmpeg -q:a 0、libvorbisにおける最高品質指定）を
// 有効な値として区別することもできるが、VideoConverter等の既存の
// 「0以下はセンチネル」規約に合わせて0もデフォルト値(4)へフォールバックする
// （既存コンストラクタ群との規約一貫性を優先した設計）。
func NewMidiConverter(
	soundfontPath string,
	sampleRate int,
	audioCodec string,
	audioQuality int,
	timeout time.Duration,
	runner CommandRunner,
) *MidiConverter {
	if soundfontPath == "" {
		soundfontPath = GetDefaultSoundfontPath()
	}
	if sampleRate <= 0 {
		sampleRate = 44100
	}
	if audioCodec == "" {
		audioCodec = "libvorbis"
	}
	if audioQuality <= 0 {
		audioQuality = 4
	}
	if timeout <= 0 {
		timeout = 300 * time.Second
	}
	if runner == nil {
		runner = NewExecCommandRunner()
	}

	return &MidiConverter{
		soundfontPath: soundfontPath,
		sampleRate:    sampleRate,
		audioCodec:    audioCodec,
		audioQuality:  audioQuality,
		timeout:       timeout,
		runner:        runner,
	}
}

// SoundfontPath は使用するサウンドフォントのパスを返す。
func (c *MidiConverter) SoundfontPath() string { return c.soundfontPath }

// SampleRate は出力サンプルレート（Hz）を返す。
func (c *MidiConverter) SampleRate() int { return c.sampleRate }

// AudioCodec は使用する音声コーデックを返す。
func (c *MidiConverter) AudioCodec() string { return c.audioCodec }

// AudioQuality は音声品質（VBR用、0-10）を返す。
func (c *MidiConverter) AudioQuality() int { return c.audioQuality }

// Timeout は変換処理のタイムアウトを返す。
func (c *MidiConverter) Timeout() time.Duration { return c.timeout }

// SupportedExtensions は対応する拡張子の一覧を返す。
func (c *MidiConverter) SupportedExtensions() []string {
	return []string{".mid", ".midi"}
}

// GetOutputExtension は変換後ファイルの拡張子".ogg"を返す。
//
// why not: Converterインターフェースが持つGetOutputExtensionは出力パスの
// 拡張子を書き換えるために使われる。MIDI変換の出力実体は常にOGGであるため、
// 空文字列（拡張子不変）を返すと出力先が".mid"のままになり誤りとなる。
func (c *MidiConverter) GetOutputExtension(_ string) string { return ".ogg" }

// CanConvert はfilePathが変換可能かを拡張子で判定する。
func (c *MidiConverter) CanConvert(filePath string) bool {
	ext := strings.ToLower(filepath.Ext(filePath))

	return containsString(c.SupportedExtensions(), ext)
}

// IsFluidsynthAvailable はfluidsynthコマンドが利用可能かを確認する。
func (c *MidiConverter) IsFluidsynthAvailable() bool {
	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
	defer cancel()

	_, err := c.runner.Run(ctx, "fluidsynth", "--version")

	return err == nil
}

// Convert はMIDIファイルをOGG Vorbis形式に変換し、destへ出力する。
//
// 実行される実効的なコマンドは以下の通り:
// `fluidsynth -ni -g 1.0 -r <sampleRate> -F <一時WAV> <soundfont> <source>`
// に続けて
// `ffmpeg -y -i <一時WAV> -c:a <audioCodec> -q:a <audioQuality> <dest>`
//
// why not: FluidSynth/FFmpegそれぞれの未インストールとタイムアウトを区別した
// 専用メッセージを返すこともできるが、CommandRunner抽象化により両者とも
// 単一のerr値に収束する（video.goのVideoConverter.Convertと同じ簡略化）。
// 区別が必要になればCommandRunnerの実装側でセンチネルエラーを定義して
// 呼び出し元に伝播させる。
func (c *MidiConverter) Convert(source, dest string) (ConversionResult, error) {
	if _, err := os.Stat(source); err != nil {
		return ConversionResult{
			SourcePath: source,
			Status:     StatusFailed,
			Message:    fmt.Sprintf("変換元ファイルが見つかりません: %s", source),
		}, nil
	}

	if _, err := os.Stat(c.soundfontPath); err != nil {
		return ConversionResult{
			SourcePath: source,
			Status:     StatusFailed,
			Message:    fmt.Sprintf("サウンドフォントが見つかりません: %s", c.soundfontPath),
		}, nil
	}

	bytesBefore := getFileSize(source)

	if err := os.MkdirAll(filepath.Dir(dest), 0o750); err != nil {
		return ConversionResult{
			SourcePath: source,
			Status:     StatusFailed,
			Message:    fmt.Sprintf("MIDI変換に失敗しました: %s", err),
		}, nil
	}

	tmpWav, err := os.CreateTemp("", "*.wav")
	if err != nil {
		return ConversionResult{
			SourcePath: source,
			Status:     StatusFailed,
			Message:    fmt.Sprintf("MIDI変換に失敗しました: %s", err),
		}, nil
	}
	tmpWavPath := tmpWav.Name()
	_ = tmpWav.Close()
	// why: FluidSynth/FFmpegどちらが失敗しても一時ファイルが残らないよう
	// deferで無条件に削除する。
	defer func() { _ = os.Remove(tmpWavPath) }()

	if result := c.runFluidsynth(source, tmpWavPath); result != nil {
		return *result, nil
	}

	if result := c.runFFmpeg(tmpWavPath, dest); result != nil {
		return *result, nil
	}

	return ConversionResult{
		SourcePath:  source,
		DestPath:    dest,
		Status:      StatusSuccess,
		BytesBefore: bytesBefore,
		BytesAfter:  getFileSize(dest),
	}, nil
}

// runFluidsynth はFluidSynthを実行してsourceをwavOutputへレンダリングする。
// エラー時は*ConversionResultを、成功時はnilを返す。
func (c *MidiConverter) runFluidsynth(source, wavOutput string) *ConversionResult {
	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
	defer cancel()

	args := []string{
		"-ni",       // No interactive mode, no shell
		"-g", "1.0", // Gain
		"-r", strconv.Itoa(c.sampleRate), // Sample rate
		"-F", wavOutput, // Output file
		c.soundfontPath,
		source,
	}

	if _, err := c.runner.Run(ctx, "fluidsynth", args...); err != nil {
		return &ConversionResult{
			SourcePath: source,
			Status:     StatusFailed,
			Message:    fmt.Sprintf("FluidSynth変換に失敗しました: %s", err),
		}
	}

	return nil
}

// runFFmpeg はFFmpegを実行してwavInputをoggOutputへ変換する。
// エラー時は*ConversionResultを、成功時はnilを返す。
func (c *MidiConverter) runFFmpeg(wavInput, oggOutput string) *ConversionResult {
	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
	defer cancel()

	args := []string{
		"-y", // Overwrite output
		"-i", wavInput,
		"-c:a", c.audioCodec,
		"-q:a", strconv.Itoa(c.audioQuality),
		oggOutput,
	}

	if _, err := c.runner.Run(ctx, "ffmpeg", args...); err != nil {
		return &ConversionResult{
			SourcePath: wavInput,
			Status:     StatusFailed,
			Message:    fmt.Sprintf("FFmpeg変換に失敗しました: %s", err),
		}
	}

	return nil
}
